Process evcxr_jupter cannot exit automatically after zed exit

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Once start REPL for rust, zed create a process called evcxr_jupter.exe.
![image](https://github.com/user-attachments/assets/ed5feb19-3428-4a8a-9549-a4e7ba9557e4)

If I just close zed without shutdowning the kernel, evcxr_jupter will not exit.  
It shows running in task manager.
![image](https://github.com/user-attachments/assets/5dc50ccc-dfea-4c9a-aeb3-138ab768bc3a).

My evcxr_jupter.exe takes the .rustup tmp up.   
When I run `rustup update` the directory of tmp cannot be cleaned up for `access denied`, which I just terminated the process and solved it.
![image](https://github.com/user-attachments/assets/0ad49726-f503-434c-8188-b26263d848ce)

### Environment

Zed: v0.154.0 (Zed Dev 444115080994413215680f5a43cdf6b5d61cc32d)
OS: Windows 10.0.19045
Memory: 15.9 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics 610 || Intel Corporation || Intel driver

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->
2024-09-18T11:54:13.0080084+08:00 [INFO] ========== starting zed ==========
2024-09-18T11:54:13.2483264+08:00 [INFO] Use Microsoft YaHei UI as UI font.
2024-09-18T11:54:13.250374+08:00 [INFO] Opening main db
2024-09-18T11:54:13.2655765+08:00 [INFO] Using git binary path: None
2024-09-18T11:54:13.2667652+08:00 [ERROR] Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-09-18T11:54:13.2669526+08:00 [ERROR] SHELL environment variable is not assigned so we can't source login environment variables

Caused by:
    environment variable not found
2024-09-18T11:54:13.3636868+08:00 [ERROR] Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-09-18T11:54:13.3660199+08:00 [INFO] initializing context server client
2024-09-18T11:54:13.3663002+08:00 [INFO] Registering built-in prompt template: content_prompt
2024-09-18T11:54:13.3672668+08:00 [INFO] Registering built-in prompt template: edit_workflow
2024-09-18T11:54:13.3677376+08:00 [INFO] Registering built-in prompt template: step_resolution
2024-09-18T11:54:13.368603+08:00 [INFO] Registering built-in prompt template: terminal_assistant_prompt
2024-09-18T11:54:13.3696946+08:00 [INFO] No prompt template overrides directory found at C:\Users\Ad\AppData\Local\Zed\prompt_overrides. Using built-in prompts.
2024-09-18T11:54:13.3978413+08:00 [INFO] extensions updated. loading 1, reloading 0, unloading 0
2024-09-18T11:54:13.8037894+08:00 [ERROR] 找不到元素。 (0x80070490)
2024-09-18T11:54:13.8177498+08:00 [INFO] Opening main db
2024-09-18T11:54:13.8211622+08:00 [ERROR] Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-09-18T11:54:13.901299+08:00 [INFO] Enabling color space support
2024-09-18T11:54:13.9231786+08:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-09-18T11:54:13.9447303+08:00 [INFO] Adapter: "Intel(R) UHD Graphics 610"
2024-09-18T11:54:13.9455036+08:00 [INFO] No ray tracing extensions are supported
2024-09-18T11:54:14.0873678+08:00 [WARN] Requested size 1x1 is outside of surface capabilities
2024-09-18T11:54:14.0875308+08:00 [ERROR] No composite alpha flag for transparency: OPAQUE | INHERIT
2024-09-18T11:54:14.0876444+08:00 [INFO] Using surface present mode IMMEDIATE
2024-09-18T11:54:14.0877641+08:00 [WARN] Unable to forbid exclusive full screen
2024-09-18T11:54:14.1610533+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-09-18T11:54:14.1613243+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-09-18T11:54:14.1615617+08:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2024-09-18T11:54:14.1654485+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1662641+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1677277+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1684614+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1701184+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1707172+08:00 [INFO] Skip function Some("to_device_position_impl")
2024-09-18T11:54:14.170842+08:00 [INFO] Skip function Some("to_device_position")
2024-09-18T11:54:14.1709554+08:00 [INFO] Skip function Some("to_device_position_transformed")
2024-09-18T11:54:14.1710737+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1713507+08:00 [INFO] Skip function Some("blend_color")
2024-09-18T11:54:14.172596+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1734654+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1740643+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1753582+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1765896+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1774049+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.1779237+08:00 [INFO] Skip function Some("to_device_position_impl")
2024-09-18T11:54:14.1780429+08:00 [INFO] Skip function Some("to_device_position")
2024-09-18T11:54:14.17815+08:00 [INFO] Skip function Some("to_device_position_transformed")
2024-09-18T11:54:14.1782558+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.178489+08:00 [INFO] Skip function Some("blend_color")
2024-09-18T11:54:14.2141195+08:00 [ERROR] No composite alpha flag for transparency: OPAQUE | INHERIT
2024-09-18T11:54:14.2142456+08:00 [INFO] Using surface present mode IMMEDIATE
2024-09-18T11:54:14.2143606+08:00 [WARN] Unable to forbid exclusive full screen
2024-09-18T11:54:14.2529654+08:00 [ERROR] No composite alpha flag for transparency: OPAQUE | INHERIT
2024-09-18T11:54:14.2530965+08:00 [INFO] Using surface present mode IMMEDIATE
2024-09-18T11:54:14.2532164+08:00 [WARN] Unable to forbid exclusive full screen
2024-09-18T11:54:14.3872842+08:00 [INFO] Using surface present mode IMMEDIATE
2024-09-18T11:54:14.387423+08:00 [WARN] Unable to forbid exclusive full screen
2024-09-18T11:54:14.42237+08:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2024-09-18T11:54:14.4259888+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4266342+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4275456+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4281828+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4291236+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4297225+08:00 [INFO] Skip function Some("to_device_position_impl")
2024-09-18T11:54:14.4298508+08:00 [INFO] Skip function Some("to_device_position")
2024-09-18T11:54:14.4299639+08:00 [INFO] Skip function Some("to_device_position_transformed")
2024-09-18T11:54:14.4300823+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4303228+08:00 [INFO] Skip function Some("blend_color")
2024-09-18T11:54:14.4314745+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4322259+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4328931+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4361396+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4382028+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.438968+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4394767+08:00 [INFO] Skip function Some("to_device_position_impl")
2024-09-18T11:54:14.4396113+08:00 [INFO] Skip function Some("to_device_position")
2024-09-18T11:54:14.4397212+08:00 [INFO] Skip function Some("to_device_position_transformed")
2024-09-18T11:54:14.4398302+08:00 [INFO] Skip function Some("to_tile_position")
2024-09-18T11:54:14.4400699+08:00 [INFO] Skip function Some("blend_color")
2024-09-18T11:54:14.4419541+08:00 [INFO] Using GPU: GPUSpecs { is_software_emulated: false, device_name: "Intel(R) UHD Graphics 610", driver_name: "Intel Corporation", driver_info: "Intel driver" }
2024-09-18T11:54:14.4427616+08:00 [ERROR] Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-09-18T11:54:14.4510757+08:00 [INFO] Opening main db
2024-09-18T11:54:14.4652799+08:00 [INFO] Opening main db
2024-09-18T11:54:14.4782025+08:00 [INFO] Node runtime install_if_needed
2024-09-18T11:54:14.490052+08:00 [INFO] Creating texture 0x27d246b30f8 of size 1024x1024x1 and format R8Unorm, name 'atlas', handle 0
2024-09-18T11:54:14.4948798+08:00 [INFO] Creating buffer 0x27d24ba10c8 of size 65536, name 'chunk-0', handle 1
2024-09-18T11:54:14.6383934+08:00 [INFO] Creating buffer 0x27d24ba1be8 of size 4096, name 'chunk-0', handle 2
2024-09-18T11:54:14.6386261+08:00 [INFO] Creating buffer 0x27d24ba2178 of size 4096, name 'chunk-1', handle 3
2024-09-18T11:54:14.7944412+08:00 [INFO] Creating buffer 0x27d24abafb8 of size 4368, name 'chunk-2', handle 4
2024-09-18T11:54:14.7946303+08:00 [INFO] Creating buffer 0x27d24ab8e58 of size 14000, name 'chunk-3', handle 5
2024-09-18T11:54:14.9425606+08:00 [INFO] Creating buffer 0x27d24abb548 of size 20944, name 'chunk-4', handle 6
2024-09-18T11:54:14.9427909+08:00 [INFO] Creating buffer 0x27d24abcb88 of size 19152, name 'chunk-5', handle 7
2024-09-18T11:54:14.949617+08:00 [INFO] attempting to start language server "rust-analyzer", path: "E:\\Projects\\rust-exercise", id: 1
2024-09-18T11:54:14.9530823+08:00 [INFO] Initializing default prettier with plugins {}
2024-09-18T11:54:14.9540739+08:00 [INFO] Initializing default prettier with plugins {}
2024-09-18T11:54:14.9634075+08:00 [INFO] found user-installed language server for rust-analyzer. path: "D:\\Scoop\\apps\\rustup\\current\\.cargo\\bin\\rust-analyzer.exe", arguments: []
2024-09-18T11:54:14.9638692+08:00 [INFO] starting language server process. binary path: "D:\\Scoop\\apps\\rustup\\current\\.cargo\\bin\\rust-analyzer.exe", working directory: "E:\\Projects\\rust-exercise", args: []
2024-09-18T11:54:15.0741753+08:00 [INFO] Creating buffer 0x27d24abdc38 of size 39536, name 'chunk-6', handle 8
2024-09-18T11:54:15.0881319+08:00 [INFO] Creating buffer 0x27d24ab6768 of size 39536, name 'chunk-7', handle 9
2024-09-18T11:54:15.0884292+08:00 [INFO] Creating buffer 0x27d24ab6cf8 of size 19152, name 'chunk-8', handle 10
2024-09-18T11:54:15.4782842+08:00 [INFO] starting language server process. binary path: "C:\\Users\\Ad\\AppData\\Local\\Zed\\node\\node-v22.5.1-win-x64\\node.exe", working directory: "C:/", args: ["C:\\Users\\Ad\\AppData\\Local\\Zed\\copilot\\copilot-v0.5.0\\dist/agent.js", "--stdio"]
2024-09-18T11:54:15.8802901+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:15.8804717+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:15.8805743+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:15.8806713+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.0653913+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.0704827+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.0706655+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.0707784+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.0708866+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.1144571+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.1146463+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.1152487+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:16.8449872+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:17.0265381+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:17.0267443+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:17.410196+08:00 [WARN] request completed with error: failed to connect to the server
2024-09-18T11:54:17.5135674+08:00 [INFO] Creating texture 0x27d246b4518 of size 1024x1024x1 and format R16Float, name 'atlas', handle 11
2024-09-18T11:54:17.5138162+08:00 [INFO] Creating buffer 0x27d24ab7818 of size 40544, name 'chunk-9', handle 12
2024-09-18T11:54:17.7342945+08:00 [INFO] Language server with id 0 sent unhandled notification LogMessage:
{
  "level": 0,
  "message": "[DEBUG] [agent] [2024-09-18T03:54:17.727Z] Agent service starting",
  "metadataStr": "[DEBUG] [agent] [2024-09-18T03:54:17.727Z]",
  "extra": [
    "Agent service starting"
  ]
}
2024-09-18T11:54:17.7768815+08:00 [INFO] Language server with id 0 sent unhandled notification client/registerCapability:
{
  "registrations": [
    {
      "id": "d6f2beff-bd29-4eb9-ba3b-5927dc995f3b",
      "method": "workspace/didChangeWorkspaceFolders",
      "registerOptions": {}
    }
  ]
}
2024-09-18T11:54:18.0058996+08:00 [INFO] Creating buffer 0x27d24abd6a8 of size 41216, name 'chunk-10', handle 13
2024-09-18T11:54:18.0063064+08:00 [INFO] Language server with id 0 sent unhandled notification LogMessage:
{
  "level": 0,
  "message": "[DEBUG] [agent] [2024-09-18T03:54:17.854Z] Telemetry initialized",
  "metadataStr": "[DEBUG] [agent] [2024-09-18T03:54:17.854Z]",
  "extra": [
    "Telemetry initialized"
  ]
}
2024-09-18T11:54:18.0200348+08:00 [INFO] Creating buffer 0x27d24ab7da8 of size 41552, name 'chunk-11', handle 14
2024-09-18T11:54:18.0276747+08:00 [INFO] Creating buffer 0x27d245dc9a8 of size 41552, name 'chunk-12', handle 15
2024-09-18T11:54:18.6785877+08:00 [INFO] Creating buffer 0x27d245dd4c8 of size 41888, name 'chunk-13', handle 16
2024-09-18T11:54:18.6787577+08:00 [INFO] Creating buffer 0x27d245de578 of size 44016, name 'chunk-14', handle 17
2024-09-18T11:54:18.6924669+08:00 [INFO] Creating buffer 0x27d245deb08 of size 46368, name 'chunk-15', handle 18
2024-09-18T11:54:20.3913728+08:00 [INFO] Creating buffer 0x27d24ba3228 of size 46928, name 'chunk-16', handle 19
2024-09-18T11:54:20.4580884+08:00 [INFO] Creating buffer 0x27d24ba5ea8 of size 47488, name 'chunk-17', handle 20
2024-09-18T11:54:21.0131152+08:00 [ERROR] Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-09-18T11:54:21.8605864+08:00 [INFO] Creating buffer 0x27d24bb63d8 of size 48720, name 'chunk-18', handle 21
2024-09-18T11:54:22.0074344+08:00 [INFO] Creating buffer 0x27d24bb6968 of size 48720, name 'chunk-19', handle 22
2024-09-18T11:54:22.2085705+08:00 [INFO] Creating buffer 0x27d24bb4808 of size 49280, name 'chunk-20', handle 23
2024-09-18T11:54:22.3125941+08:00 [INFO] Initializing default prettier with plugins {}
2024-09-18T11:54:25.6622565+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-09-18T11:54:25.6733891+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-09-18T11:54:25.6911551+08:00 [INFO] Creating a descriptor pool for at most 256 sets
2024-09-18T11:54:25.7077384+08:00 [INFO] Creating a descriptor pool for at most 256 sets
2024-09-18T11:54:49.751366+08:00 [ERROR] Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-09-18T11:54:49.927567+08:00 [WARN] Generic lsp request to rust-analyzer failed: content modified
2024-09-18T11:54:49.9277628+08:00 [WARN] Generic lsp request to rust-analyzer failed: content modified
2024-09-18T11:54:49.9280216+08:00 [ERROR] content modified
2024-09-18T11:54:49.9281541+08:00 [ERROR] content modified
2024-09-18T11:55:16.6225008+08:00 [INFO] Destroying texture 0x27d246b30f8, handle 0
2024-09-18T11:55:16.6227858+08:00 [INFO] Destroying texture 0x27d246b4518, handle 11
2024-09-18T11:55:16.623323+08:00 [INFO] Destroying buffer 0x27d24ba10c8, handle 1
2024-09-18T11:55:16.62379+08:00 [INFO] Destroying buffer 0x27d24abb548, handle 6
2024-09-18T11:55:16.6239284+08:00 [INFO] Destroying buffer 0x27d245dc9a8, handle 15
2024-09-18T11:55:16.6240478+08:00 [INFO] Destroying buffer 0x27d24abdc38, handle 8
2024-09-18T11:55:16.6241585+08:00 [INFO] Destroying buffer 0x27d24ba1be8, handle 2
2024-09-18T11:55:16.6242676+08:00 [INFO] Destroying buffer 0x27d245de578, handle 17
2024-09-18T11:55:16.624378+08:00 [INFO] Destroying buffer 0x27d24ba5ea8, handle 20
2024-09-18T11:55:16.624487+08:00 [INFO] Destroying buffer 0x27d24ab8e58, handle 5
2024-09-18T11:55:16.6245977+08:00 [INFO] Destroying buffer 0x27d245dd4c8, handle 16
2024-09-18T11:55:16.6247362+08:00 [INFO] Destroying buffer 0x27d24bb6968, handle 22
2024-09-18T11:55:16.6248607+08:00 [INFO] Destroying buffer 0x27d24ab6cf8, handle 10
2024-09-18T11:55:16.6249953+08:00 [INFO] Destroying buffer 0x27d245deb08, handle 18
2024-09-18T11:55:16.6251199+08:00 [INFO] Destroying buffer 0x27d24ab6768, handle 9
2024-09-18T11:55:16.6252769+08:00 [INFO] Destroying buffer 0x27d24ba2178, handle 3
2024-09-18T11:55:16.625461+08:00 [INFO] Destroying buffer 0x27d24ab7da8, handle 14
2024-09-18T11:55:16.6255671+08:00 [INFO] Destroying buffer 0x27d24bb63d8, handle 21
2024-09-18T11:55:16.6256833+08:00 [INFO] Destroying buffer 0x27d24abafb8, handle 4
2024-09-18T11:55:16.6257871+08:00 [INFO] Destroying buffer 0x27d24ba3228, handle 19
2024-09-18T11:55:16.6258938+08:00 [INFO] Destroying buffer 0x27d24bb4808, handle 23
2024-09-18T11:55:16.6259997+08:00 [INFO] Destroying buffer 0x27d24abcb88, handle 7
2024-09-18T11:55:16.6261052+08:00 [INFO] Destroying buffer 0x27d24abd6a8, handle 13
2024-09-18T11:55:16.6262082+08:00 [INFO] Destroying buffer 0x27d24ab7818, handle 12
2024-09-18T11:55:16.697712+08:00 [ERROR] window not found
<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

--- a/crates/repl/src/kernels/mod.rs
+++ b/crates/repl/src/kernels/mod.rs
@@ -170,6 +170,7 @@ pub trait RunningKernel: Send + Debug {
     fn kernel_info(&self) -> Option<&KernelInfoReply>;
     fn set_kernel_info(&mut self, info: KernelInfoReply);
     fn force_shutdown(&mut self, window: &mut Window, cx: &mut App) -> Task<anyhow::Result<()>>;
+    fn kill(&mut self);
 }
 
 #[derive(Debug, Clone)]

--- a/crates/repl/src/kernels/native_kernel.rs
+++ b/crates/repl/src/kernels/native_kernel.rs
@@ -370,17 +370,21 @@ impl RunningKernel for NativeRunningKernel {
     }
 
     fn force_shutdown(&mut self, _window: &mut Window, _cx: &mut App) -> Task<anyhow::Result<()>> {
+        self.kill();
+        Task::ready(Ok(()))
+    }
+
+    fn kill(&mut self) {
         self._process_status_task.take();
         self.request_tx.close_channel();
-        Task::ready(self.process.kill().context("killing the kernel process"))
+        self.process.kill().ok();
     }
 }
 
 impl Drop for NativeRunningKernel {
     fn drop(&mut self) {
         std::fs::remove_file(&self.connection_path).ok();
-        self.request_tx.close_channel();
-        self.process.kill().ok();
+        self.kill();
     }
 }
 

--- a/crates/repl/src/kernels/remote_kernels.rs
+++ b/crates/repl/src/kernels/remote_kernels.rs
@@ -289,4 +289,8 @@ impl RunningKernel for RemoteRunningKernel {
             Ok(())
         })
     }
+
+    fn kill(&mut self) {
+        self.request_tx.close_channel();
+    }
 }

--- a/crates/repl/src/repl_store.rs
+++ b/crates/repl/src/repl_store.rs
@@ -289,7 +289,11 @@ impl ReplStore {
     fn shutdown_all_sessions(&mut self, cx: &mut Context<Self>) -> impl Future<Output = ()> + use<> {
         for session in self.sessions.values() {
             session.update(cx, |session, _cx| {
-                session.kernel = Kernel::Shutdown;
+                if let Kernel::RunningKernel(mut kernel) =
+                    std::mem::replace(&mut session.kernel, Kernel::Shutdown)
+                {
+                    kernel.kill();
+                }
             });
         }
         self.sessions.clear();

--- a/crates/repl/src/repl_store.rs
+++ b/crates/repl/src/repl_store.rs
@@ -11,8 +11,7 @@ use project::{Fs, Project, WorktreeId};
 use settings::{Settings, SettingsStore};
 
 use crate::kernels::{
-    Kernel, list_remote_kernelspecs, local_kernel_specifications,
-    python_env_kernel_specifications,
+    Kernel, list_remote_kernelspecs, local_kernel_specifications, python_env_kernel_specifications,
 };
 use crate::{JupyterSettings, KernelSpecification, Session};
 
@@ -286,7 +285,10 @@ impl ReplStore {
         self.sessions.remove(&entity_id);
     }
 
-    fn shutdown_all_sessions(&mut self, cx: &mut Context<Self>) -> impl Future<Output = ()> + use<> {
+    fn shutdown_all_sessions(
+        &mut self,
+        cx: &mut Context<Self>,
+    ) -> impl Future<Output = ()> + use<> {
         for session in self.sessions.values() {
             session.update(cx, |session, _cx| {
                 if let Kernel::RunningKernel(mut kernel) =

--- a/crates/repl/src/repl_store.rs
+++ b/crates/repl/src/repl_store.rs
@@ -1,3 +1,4 @@
+use std::future::Future;
 use std::sync::Arc;
 
 use anyhow::{Context as _, Result};
@@ -10,7 +11,8 @@ use project::{Fs, Project, WorktreeId};
 use settings::{Settings, SettingsStore};
 
 use crate::kernels::{
-    list_remote_kernelspecs, local_kernel_specifications, python_env_kernel_specifications,
+    Kernel, list_remote_kernelspecs, local_kernel_specifications,
+    python_env_kernel_specifications,
 };
 use crate::{JupyterSettings, KernelSpecification, Session};
 
@@ -47,9 +49,12 @@ impl ReplStore {
     }
 
     pub fn new(fs: Arc<dyn Fs>, cx: &mut Context<Self>) -> Self {
-        let subscriptions = vec![cx.observe_global::<SettingsStore>(move |this, cx| {
-            this.set_enabled(JupyterSettings::enabled(cx), cx);
-        })];
+        let subscriptions = vec![
+            cx.observe_global::<SettingsStore>(move |this, cx| {
+                this.set_enabled(JupyterSettings::enabled(cx), cx);
+            }),
+            cx.on_app_quit(Self::shutdown_all_sessions),
+        ];
 
         let this = Self {
             fs,
@@ -279,6 +284,16 @@ impl ReplStore {
 
     pub fn remove_session(&mut self, entity_id: EntityId) {
         self.sessions.remove(&entity_id);
+    }
+
+    fn shutdown_all_sessions(&mut self, cx: &mut Context<Self>) -> impl Future<Output = ()> + use<> {
+        for session in self.sessions.values() {
+            session.update(cx, |session, _cx| {
+                session.kernel = Kernel::Shutdown;
+            });
+        }
+        self.sessions.clear();
+        futures::future::ready(())
     }
 
     #[cfg(test)]

